### PR TITLE
[Attributor] Add support for atomic operations in `AAAddressSpace`

### PR DESCRIPTION
**Pull request by shiltian**

### CODE REVIEW

**Comment by shiltian** [comment]
* **#106927** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/106927?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @shiltian and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/lib/Transforms/IPO/AttributorAttributes.cpp [hunk 1]
```diff
@@ -12583,10 +12583,14 @@ struct AAAddressSpaceImpl : public AAAddressSpace {
...
       if (isa<LoadInst>(Inst))
         MakeChange(Inst, const_cast<Use &>(U));
-      if (isa<StoreInst>(Inst)) {
+      else if (isa<StoreInst>(Inst)) {
         // We only make changes if the use is the pointer operand.
         if (U.getOperandNo() == 1)
           MakeChange(Inst, const_cast<Use &>(U));
+      } else if (isa<AtomicRMWInst>(Inst) || isa<AtomicCmpXchgInst>(Inst)) {
+        // We only make changes if the use is the pointer operand.
+        if (U.getOperandNo() == 0)
```
Comment (arsenm) [inline llvm/lib/Transforms/IPO/AttributorAttributes.cpp:12592]: This has the same bug that load and store have above, where it doesn't skip volatiles. InferAddressSpaces skips volatile && !TTI.hasVolatileVariant

Comment (arsenm) [inline llvm/lib/Transforms/IPO/AttributorAttributes.cpp:12592]: Hardcoding the pointer operand index, use the AtoimcRMWInst::getPointerOperandIndex

Comment (shiltian) [reply llvm/lib/Transforms/IPO/AttributorAttributes.cpp:12592]: https://github.com/llvm/llvm-project/pull/107250

File: llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp [hunk 2]
```diff
@@ -1084,10 +1084,15 @@ static bool runImpl(Module &M, AnalysisGetter &AG, TargetMachine &TM,
...
-      if (auto *SI = dyn_cast<StoreInst>(&I)) {
+      } else if (auto *SI = dyn_cast<StoreInst>(&I)) {
         A.getOrCreateAAFor<AAAddressSpace>(
             IRPosition::value(*SI->getPointerOperand()));
+      } else if (auto *AI = dyn_cast<AtomicRMWInst>(&I)) {
+        A.getOrCreateAAFor<AAAddressSpace>(
+            IRPosition::value(*AI->getPointerOperand()));
+      } else if (auto *CmpX = dyn_cast<AtomicCmpXchgInst>(&I)) {
+        A.getOrCreateAAFor<AAAddressSpace>(
+            IRPosition::value(*CmpX->getPointerOperand()));
```
Comment (ssahasra) [inline llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:1095]: This a non-blocking request, but all that copy-pasted code is really itchy. Would you consider adding ``getPointerOperand()`` to the ``Instruction`` class itself?

Comment (shiltian) [reply llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:1095]: This is handled in https://github.com/llvm/llvm-project/pull/107250. Will rebase this one after https://github.com/llvm/llvm-project/pull/107250 is landed.

File: llvm/test/CodeGen/AMDGPU/aa-as-infer.ll [hunk 3]
```diff
@@ -86,3 +86,163 @@ define void @call_volatile_load_store_as_4(ptr addrspace(4) %p1, ptr addrspace(4
...
+; CHECK-NEXT:    [[CMPXCHG_1:%.*]] = cmpxchg ptr addrspace(1) [[TMP2]], i32 0, i32 5 acq_rel monotonic, align 4
+; CHECK-NEXT:    [[TMP3:%.*]] = addrspacecast ptr [[WORD]] to ptr addrspace(1)
+; CHECK-NEXT:    [[CMPXCHG_2:%.*]] = cmpxchg ptr addrspace(1) [[TMP3]], i32 0, i32 6 acquire monotonic, align 4
+; CHECK-NEXT:    [[TMP4:%.*]] = addrspacecast ptr [[WORD]] to ptr addrspace(1)
+; CHECK-NEXT:    [[CMPXCHG_3:%.*]] = cmpxchg ptr addrspace(1) [[TMP4]], i32 0, i32 7 release monotonic, align 4
+; CHECK-NEXT:    [[TMP5:%.*]] = addrspacecast ptr [[WORD]] to ptr addrspace(1)
+; CHECK-NEXT:    [[CMPXCHG_4:%.*]] = cmpxchg ptr addrspace(1) [[TMP5]], i32 0, i32 8 seq_cst monotonic, align 4
+; CHECK-NEXT:    [[TMP6:%.*]] = addrspacecast ptr [[WORD]] to ptr addrspace(1)
+; CHECK-NEXT:    [[CMPXCHG_5:%.*]] = cmpxchg weak ptr addrspace(1) [[TMP6]], i32 0, i32 9 seq_cst monotonic, align 4
+; CHECK-NEXT:    [[TMP7:%.*]] = addrspacecast ptr [[WORD]] to ptr addrspace(1)
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/aa-as-infer.ll:105]: It's not ideal this ends up creating the same addrspacecast for every instance

**Review by arsenm — APPROVED** [review]

File: llvm/test/CodeGen/AMDGPU/aa-as-infer.ll (hunk 3 above)
Comment (shiltian) [reply llvm/test/CodeGen/AMDGPU/aa-as-infer.ll:105]: yeah, need a follow-up patch to optimize this

**Pull request merged**